File: 529-minesweeper/ken/529.py
class Solution(object):
    def updateBoard(self, board, click):
        """
        :type board: List[List[str]]
        :type click: List[int]
        :rtype: List[List[str]]
        """

        height = len(board)
        width = len(board[0])
        genNeighbors = lambda click : [(row1, col1) for row1 in range(click[0]-1, click[0]+2) \
                                                    for col1 in range(click[1]-1, click[1]+2) \
                                                    if (row1 >= 0 and row1 < height and \
                                                        col1 >= 0 and col1 < width)]

        def getNewCell(board, neightbors):
            mineCnt = 0
            for neighbor in neightbors:
                if board[neighbor[0]][neighbor[1]] == 'M':
                    mineCnt += 1

            return 'B' if (mineCnt == 0) else str(mineCnt)

        def rec(board, click):
            if board[click[0]][click[1]] != 'E':
                return

            neightbors = genNeighbors(click)

            board[click[0]][click[1]] = getNewCell(board, neightbors)
            if board[click[0]][click[1]] != 'B':
                return

            for neighbor in neightbors:
                rec(board, neighbor)

        if board[click[0]][click[1]] == 'M':
            board[click[0]][click[1]] = 'X'
            return board

        rec(board, click)

        return board

File: 529-minesweeper/leochang/529.cpp
class Solution {
public:
    vector<vector<char>> updateBoard(vector<vector<char>>& board, vector<int>& click) {
        if(board[click[0]][click[1]] == 'M'){
            board[click[0]][click[1]] = 'X';
            return board;
        }
        dfs(board, click[0], click[1]);
        return board;
    }
    void dfs(vector<vector<char>>& board, int x, int y) {       
        int count = 0;
        if (board[x][y] == 'E') {
            for (int i = -1; i <= 1; i++) {
                for (int j = -1; j <= 1; j++) {
                    if (i == 0 && j == 0) {
                        continue;
                    }
                    if (x+i < 0 || x+i >= board.size() ||
                       y+j < 0 || y+j >= board[0].size()) {
                        continue;
                    }
                    if (board[x+i][y+j] == 'M' || board[x+i][y+j] == 'X') {
                        count++;
                    }
                }
            }
        
            if (count > 0) {
                board[x][y] = '0' + count;
            } else {
                board[x][y] = 'B';         
                for (int i = -1; i <= 1; i++) {
                    for (int j = -1; j <= 1; j++) {
                        if (i == 0 && j == 0) {
                            continue;
                        }
                        if (x+i < 0 || x+i >= board.size() ||
                           y+j < 0 || y+j >= board[0].size()) {
                            continue;
                        }
                        dfs(board, x+i, y+j);
                    }
                }
            }
        }
    }
};
